Have isPowerOfTwo() call std::has_single_bit() and move it to MathExtras.h

cdumez opened the pull request:
#### 209ba9726d1063586de06057455d882efa0d3480
<pre>
Have isPowerOfTwo() call std::has_single_bit() and move it to MathExtras.h
<a href="https://bugs.webkit.org/show_bug.cgi?id=292572">https://bugs.webkit.org/show_bug.cgi?id=292572</a>

Reviewed by Ryosuke Niwa.

* Source/JavaScriptCore/heap/FreeList.h:
(JSC::FreeCell::descramble):
* Source/JavaScriptCore/wasm/WasmBBQJIT64.h:
(JSC::Wasm::BBQJITImpl::BBQJIT::emitModOrDiv):
* Source/WTF/wtf/MathExtras.h:
(WTF::isPowerOfTwo):
(WTF::roundUpToMultipleOfImpl):
(WTF::roundUpToMultipleOf):
(WTF::roundUpToMultipleOfNonPowerOfTwo):
(WTF::roundDownToMultipleOf):
* Source/WTF/wtf/PageBlock.h:
* Source/WTF/wtf/StdLibExtras.h:
(WTF::isPowerOfTwo): Deleted.
(WTF::roundUpToMultipleOfImpl): Deleted.
(WTF::roundUpToMultipleOf): Deleted.
(WTF::roundUpToMultipleOfNonPowerOfTwo): Deleted.
(WTF::roundDownToMultipleOf): Deleted.
* Source/WebCore/platform/audio/cocoa/CARingBuffer.cpp:
(WebCore::CARingBuffer::CARingBuffer):

Canonical link: <a href="https://commits.webkit.org/294562@main">https://commits.webkit.org/294562@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c1828d1e0480f0d6e5ccc8ed1152ac54a82ffea8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102223 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21891 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12206 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107382 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52859 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104262 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22199 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30398 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77796 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34781 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105229 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17177 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92277 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58132 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17007 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10305 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52217 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/94894 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86846 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10375 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109758 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/100832 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29355 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21640 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86774 "Found 1 new test failure: http/tests/inspector/network/har/har-page.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29718 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88480 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86365 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21980 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31167 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8878 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/23597 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29283 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34578 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/124458 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29094 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34559 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32417 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30653 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WTF/wtf/MathExtras.h:
```diff
@@ -771,6 +776,73 @@ inline bool isRepresentableAs(double value)
+}
+
+template<typename T>
+inline constexpr T roundUpToMultipleOfNonPowerOfTwo(size_t divisor, T x)
```
These function names are super confusing! We should probably rename these functions (in a separate PR).
Maybe `roundUpToMultipleOfPowerOfTwo` and `roundUpToMultipleOf`.

cdumez committed:
Have isPowerOfTwo() call std::has_single_bit() and move it to MathExtras.h
https://bugs.webkit.org/show_bug.cgi?id=292572

Reviewed by Ryosuke Niwa.

* Source/JavaScriptCore/heap/FreeList.h:
(JSC::FreeCell::descramble):
* Source/JavaScriptCore/wasm/WasmBBQJIT64.h:
(JSC::Wasm::BBQJITImpl::BBQJIT::emitModOrDiv):
* Source/WTF/wtf/MathExtras.h:
(WTF::isPowerOfTwo):
(WTF::roundUpToMultipleOfImpl):
(WTF::roundUpToMultipleOf):
(WTF::roundUpToMultipleOfNonPowerOfTwo):
(WTF::roundDownToMultipleOf):
* Source/WTF/wtf/PageBlock.h:
* Source/WTF/wtf/StdLibExtras.h:
(WTF::isPowerOfTwo): Deleted.
(WTF::roundUpToMultipleOfImpl): Deleted.
(WTF::roundUpToMultipleOf): Deleted.
(WTF::roundUpToMultipleOfNonPowerOfTwo): Deleted.
(WTF::roundDownToMultipleOf): Deleted.
* Source/WebCore/platform/audio/cocoa/CARingBuffer.cpp:
(WebCore::CARingBuffer::CARingBuffer):

Canonical link: https://commits.webkit.org/294562@main
